src/agent/Core/Controller/CheckoutSession.cpp

```diff
@@ -178,12 +179,53 @@ Controller::maybeSend100Continue(Client *client, Request *req) {
 	}
 }
 
+typedef boost::tuple<Controller*, Client*, Request*> ControllerFinishInitiatingSessionCallbackPointers;
+
+void
+Controller::onWritable(EV_P_ ev_io *io, int revents) {
+        ControllerFinishInitiatingSessionCallbackPointers *cbs = static_cast<ControllerFinishInitiatingSessionCallbackPointers *>(io->data);
+	    Controller* that = cbs->get<0>();
+		Request * req = cbs->get<2>();
+        ev_io_stop(that->getLoop(), io);
+
+		int connect_error = 0;
+		socklen_t connect_error_len = sizeof(connect_error);
+		if (-1 == getsockopt(req->session->fd(), SOL_SOCKET, SO_ERROR, &connect_error, &connect_error_len)) {
+			int err = errno;
+			throw SystemException("Cannot check socket status", err);
+		} else if (connect_error != 0) {
+			// connect_error uses the same error codes as errno
+			throw SystemException("Cannot connect socket", connect_error);
```

Comment:
We're in an event loop callback. So rather than throwing an exception, any errors we counter should either result in:

1. a call to disconnectWithError() (if it's a permanent error), or,
2. retrying the session initiation, like what you see in initiateSession's exception handler.